src/rascal/models/numerical_kernel_gradients.hh

```diff
@@ -0,0 +1,111 @@
+/**
+ * @file   rascal/models/numerical_kernel_gradients.hh
+ *
+ * @author Felix Musil <felix.musil@epfl.ch>
+ *
+ * @date   9 April 2020
+ *
+ * @brief Implementation of numerical evaluation of kernel gradients
+ *
+ * Copyright 2019 Felix Musil COSMO (EPFL), LAMMM (EPFL)
+ *
+ * Rascal is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 3, or (at
+ * your option) any later version.
+ *
+ * Rascal is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this software; see the file LICENSE. If not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ */
+
+#ifndef SRC_RASCAL_MODELS_NUMERICAL_KERNEL_GRADIENTS_HH_
+#define SRC_RASCAL_MODELS_NUMERICAL_KERNEL_GRADIENTS_HH_
+
+#include "rascal/math/utils.hh"
+#include "rascal/models/kernels.hh"
+#include "rascal/models/sparse_kernels.hh"
+#include "rascal/structure_managers/atomic_structure.hh"
+#include "rascal/structure_managers/make_structure_manager.hh"
+
+namespace rascal {
+
+  template <class KernelImpl, class Calculator, class Manager,
+            class SparsePoints, typename Derived>
+  math::Matrix_t compute_displaced_kernel(
+      KernelImpl & kernel, Calculator & calculator, Manager & manager,
+      const SparsePoints & sparse_points, const size_t & i_atom,
+      const Eigen::MatrixBase<Derived> & disp) {
+    // get a copy of the atomic_structure object
+    auto manager_root = extract_underlying_manager<0>(manager);
+    json structure_copy = manager_root->get_atomic_structure();
+    auto atomic_structure =
+        structure_copy.template get<AtomicStructure<ThreeD>>();
+    atomic_structure.displace_position(i_atom, disp);
+    // make sure all atoms are in the unit cell
+    atomic_structure.wrap();
+    manager->update(atomic_structure);
+    calculator.compute(manager);
+    std::vector<std::remove_const_t<Manager>> managers{};
+    managers.emplace_back(manager);
+    math::Matrix_t KNM = kernel.compute(calculator, managers, sparse_points);
+    // reset neighborlist to the original structure
+    manager->update(structure_copy.template get<AtomicStructure<ThreeD>>());
+    return KNM;
+  }
+
+  template <class KernelImpl, class Calculator, class Manager,
+            class SparsePoints>
+  math::Matrix_t compute_numerical_kernel_gradient(
```

Comment:
Needs documentation (this function and the one below). Please explain that these utilities are intended for finite-difference numerical computation of the gradient of a kernel-based model, for the case where analytical gradients are not available. A mention of this functionality in the developer doc pages would also be most helpful.

Reply:
how is it now ?

Reply:
Better. I'd say "gradient... for a _single_ atomic structure" to make it clearer how this function differs from the one below, and put in addition to just `@see the other function` a line for the `@param manager` that makes this function different from the multi-manager one.

Could you add a mention of this functionality to the developer docs, e.g. [here](https://cosmo-epfl.github.io/librascal/dev_guide/how_to_add_representation.html#implement-and-test-gradients) or [here](https://cosmo-epfl.github.io/librascal/dev_guide/how_to_test.html#testing-gradients)? We'll also need to work on updating the developer docs to reflect the existence of kernel-based models (the "Models" section is currently a stub).